File: main_server/src/entities/booking/booking.cpp
#include "booking.hpp"

Booking::Booking(){}
Booking::Booking(int id) : id(id) {}
Booking::Booking(int id, std::string datetime_start, std::string datetime_end, std::string entry_time, std::string exit_time,
                  float amount, int id_user, int id_vehicle, int id_parking_slot, std::string note) :
                                                              id(id),
                                                              datetime_start(datetime_start),
                                                              datetime_end(datetime_end),
                                                              entry_time(entry_time),
                                                              exit_time(exit_time),
                                                              amount(amount),
                                                              id_user(id_user),
                                                              id_vehicle(id_vehicle),
                                                              id_parking_slot(id_parking_slot),
                                                              note(note)
                                                              {}

void Booking::setId(int id){
    this->id = id;
}

void Booking::setEntryTime(std::string entry_time){
    this->entry_time = entry_time;
}

void Booking::setExitTime(std::string exit_time){
    this->exit_time = exit_time;
}

void Booking::setAmount(float amount){
    this->amount = amount;
}

void Booking::setNote(std::string note) {
    this->note = note;
}


int Booking::getId() const{
    return id;
}

std::string Booking::getDateTimeStart() const {
    return datetime_start;
}

std::string Booking::getDateTimeEnd() const {
    return datetime_end;
}

std::string Booking::getEntryTime() const {
    return entry_time;
}

std::string Booking::getExitTime() const {
    return exit_time;
}

float Booking::getAmount() const {
    return amount;
}

int Booking::getIdUser() const {
    return id_user;
}

int Booking::getIdVehicle() const {
    return id_vehicle;
}

int Booking::getIdParkingSlot() const {
    return id_parking_slot;
}

std::string Booking::getNote() const {
    return note;
}


bool operator== (const Booking &b1, const Booking &b2)
{
    return b1.id == b2.id;
}


std::ostream& operator<<(std::ostream& os, const Booking& b)
{
    os << "Booking - Id: " << b.id << ", Datetime start: " << b.datetime_start << ", Datetime end: " << b.datetime_end
        << ", Entry time: " << b.entry_time << ", Exit time: " << b.exit_time << ", Amount: " << b.amount << ", Id User: "
        << b.id_user << ", Id Vehicle: " << b.id_vehicle << ", Id Parking Slot: " << b.id_parking_slot << ", Note: "
        << b.note << std::endl;
    return os;
}

BookingException::BookingException(const std::string & message):_message(message) {}
const char * BookingException::what() const throw() {
      return _message.c_str();
}

File: uPark_client/entities/booking.py
class Booking:
    def __init__(self, id = None, datetime_start = None, datetime_end = None, entry_time = None, exit_time = None, amount = None, id_user = None, id_vehicle = None, id_parking_slot = None, note = ""):
        self._id = id
        self._datetime_start = datetime_start
        self._datetime_end = datetime_end
        self._entry_time = entry_time
        self._exit_time = exit_time
        self._amount = amount
        self._id_user = id_user
        self._id_vehicle = id_vehicle
        self._id_parking_slot = id_parking_slot
        self._note = note


    def set_id(self, id):
        self._id = id

    def set_datetime_start(self, datetime_start):
        self._datetime_start = datetime_start

    def set_datetime_end(self, datetime_end):
        self._datetime_end = datetime_end

    def set_entry_time(self, entry_time):
        self._entry_time = entry_time

    def set_exit_time(self, exit_time):
        self._exit_time = exit_time

    # def set_amount(self, amount):
    #     self._amount = amount

    # def set_id_user(self, id_user):
    #     self._id_user = id_user
    #
    # def set_id_vehicle(self, id_vehicle):
    #     self._id_vehicle = id_vehicle
    #
    # def set_id_parking_slot(self, id_parking_slot):
    #     self._id_parking_slot = id_parking_slot

    def set_note(self, note):
        self._note = note


    def get_id(self):
        return self._id

    def get_datetime_start(self):
        return self._datetime_start

    def get_datetime_end(self):
        return self._datetime_end

    def get_entry_time(self):
        return self._entry_time

    def get_exit_time(self):
        return self._exit_time

    def get_amount(self):
        return self._amount

    def get_id_user(self):
        return self._id_user

    def get_id_vehicle(self):
        return self._id_vehicle

    def get_id_parking_slot(self):
        return self._id_parking_slot

    def get_note(self):
        return self._note

    def __eq__(self, other):
        if isinstance(other, Booking):
            return self._id == other._id
        return NotImplemented

    def __str__(self):
        booking_info = (f"Booking Id: {self._id}, datetime start: {self._datetime_start}, datetime end: {self._datetime_end}, entry time: {self._entry_time}, "
        f"exit time: {self._exit_time}, amount: {self._amount}, id user: {self._id_user}, id vehicle: {self._id_vehicle}, "
        f"id parking slot: {self._id_parking_slot}, note: {self._note}")
        return booking_info
